src/pubsub.js

```diff
@@ -0,0 +1,573 @@
+        common.setup((err, factory) => {
+          if (err) {
+            return done(err)
+          }
+
+          series([
+            (cb) => spawnWithId(factory, cb),
+            (cb) => spawnWithId(factory, cb),
+            (cb) => spawnWithId(factory, cb)
+          ], (err, nodes) => {
+            if (err) {
+              return done(err)
+            }
+
+            ipfs1 = nodes[0]
+            ipfs2 = nodes[1]
+            ipfs3 = nodes[2]
+            done()
+          })
+        })
+      })
+
+      after((done) => {
+        common.teardown(done)
+      })
+
+      describe('single node', () => {
+        describe('.publish', () => {
+          it('message from string', (done) => {
+            ipfs1.pubsub.publish(topic, 'hello friend', done)
+          })
+
+          it('message from buffer', (done) => {
+            ipfs1.pubsub.publish(topic, new Buffer('hello friend'), done)
+          })
+        })
+
+        describe('.subscribe', () => {
+          it('to one topic', (done) => {
+            const check = makeCheck(2, done)
+
+            const handler = (msg) => {
+              expect(msg.data.toString()).to.equal('hi')
+              expect(msg).to.have.property('seqno')
+              expect(Buffer.isBuffer(msg.seqno)).to.be.eql(true)
+              expect(msg).to.have.property('topicCIDs').eql([topic])
+              // TODO: broken https://github.com/ipfs/go-ipfs/issues/3522
+              // expect(msg).to.have.property('from', ipfs1.peerId.id)
+
+              ipfs1.pubsub.unsubscribe(topic, handler)
+
+              ipfs1.pubsub.ls((err, topics) => {
+                expect(err).to.not.exist
+                expect(topics).to.be.empty
+                check()
+              })
+            }
+
+            ipfs1.pubsub.subscribe(topic, handler, (err) => {
+              expect(err).to.not.exist
+              ipfs1.pubsub.publish(topic, 'hi', check)
+            })
+          })
+
+          it('attaches multiple event listeners', (done) => {
+            const check = makeCheck(3, done)
+            const handler1 = (msg) => {
+              expect(msg.data.toString()).to.be.eql('hello')
+
+              ipfs1.pubsub.unsubscribe(topic, handler1)
+
+              series([
+                (cb) => ipfs1.pubsub.ls(cb),
+                (cb) => {
+                  ipfs1.pubsub.unsubscribe(topic, handler2)
+                  cb()
+                },
+                (cb) => ipfs1.pubsub.ls(cb)
+              ], (err, res) => {
+                expect(err).to.not.exist
+
+                // Still subscribed as there is one listener left
+                expect(res[0]).to.be.eql([topic])
+                // Now all listeners are gone no subscription anymore
+                expect(res[2]).to.be.eql([])
+                check()
+              })
+            }
+
+            const handler2 = (msg) => {
+              expect(msg.data.toString()).to.be.eql('hello')
+              check()
+            }
```

Comment:
You have used `function` I try to use `const name = ()` especially inside other functions and to avoid using functions before they are declared, as it is very unfortunate that `function` declarations are hoisted to the top level, allowing for code to use functions before they have been declared.